Sprint 17

SnakeAssassin opened the pull request:
Выполнены все требования согласно заданию спринта 17
* Реализовано контекстное меню для треков на главном экране
* Реализованы фильтры
* Выполнены скриншот-тесты для главного экрана,
* Реализована тёмная тема,
* Подключена AppMetrica с помощью CocoaPods и присылаются события для действий на главном экране,
* Реализован экран статистики,
* Есть английская локализация

Commit:
Добавлена RU/ENG локализация, литералы заменены на ключи из localizable

Commit:
Добавлена RU/ENG локализация. Все литералы заменены на ключи из localized

Commit:
Исправлена ошибка постоянного показа онбординг экрана при запуске приложения.

Commit:
Спринт 17 все изменения
* Реализовано контекстное меню для треков на главном экране
* Реализованы фильтры
* Выполнены скриншот-тесты для главного экрана,
* Реализована тёмная тема,
* Подключена AppMetrica с помощью CocoaPods и присылаются события для действий на главном экране,
* Реализован экран статистики,
* Есть английская локализация

andrew-kuzn-yndx commented on CoreData/UIColorMarshalling.swift:
```diff
@@ -1,27 +1,27 @@
+//            alpha: CGFloat(1.0)
+//        )
+//    }
+//}
```
Старайся не оставлять такое, правилом хорошего тона является чистить код от лишнего перед сдачей работы. В основную ветку проекта должен попадать только чистый код без лишних принтов и закомментированных блоков)

SnakeAssassin replied:
> Старайся не оставлять такое, правилом хорошего тона является чистить код от лишнего перед сдачей работы. В основную ветку проекта должен попадать только чистый код без лишних принтов и закомментированных блоков)

Странно.. У меня этого файла нет в проекте. Но на диске в папке проекта есть. Как в Xcode увидеть скрытые удаленные файлы?

andrew-kuzn-yndx replied:
Да, в проекте их может не быть, но они могут все ещё лежать в папках проекта 
Нужно найти соответствующую папку в проводнике, файл может лежать там)

andrew-kuzn-yndx commented on Tracker/Controllers/AddTrackerViewController/AddTrackerViewController.swift:
```diff
@@ -1,22 +1,20 @@
     weak var delegate: AddTrackerViewControllerDelegate?
     
-    // MARK: Private Properties
+    //MARK: View
```
Молодец, что используешь MARK!

andrew-kuzn-yndx commented on Tracker/Controllers/CategoryViewController/CategoryListViewController.swift:
```diff
@@ -0,0 +1,275 @@
+        imageView.translatesAutoresizingMaskIntoConstraints = false
+        return imageView
+    }()
+    private lazy var stubLabel: UILabel = {
```
Немного плотно) блоки кода можно разделить пустой строкой

andrew-kuzn-yndx commented on Tracker/Controllers/CategoryViewController/CategoryListViewController.swift:
```diff
@@ -0,0 +1,275 @@
+        } else {
+            listTableView.isHidden = false
+            stubLabel.isHidden = true
+            stubImageView.isHidden = true
```
Тут можно упростить и сократить, например:
let isContainCategories = viewModel.conditionStubs()
listTableView.isHidden = isContainCategories
            stubLabel.isHidden = !isContainCategories
            stubImageView.isHidden = !isContainCategories

andrew-kuzn-yndx commented on Tracker/Controllers/CategoryViewController/CategoryListViewController.swift:
```diff
@@ -0,0 +1,275 @@
+    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
+        let cell = tableView.dequeueReusableCell(withIdentifier: "cell", for: indexPath)
+        cell.selectionStyle = .none
+        cell.backgroundColor = .ypLigthGray.withAlphaComponent(0.3)
```
Эти свойства лучше задавать внутри ячейки

andrew-kuzn-yndx commented on Tracker/Controllers/CategoryViewController/CategoryListViewController.swift:
```diff
@@ -0,0 +1,275 @@
+        let cell = tableView.dequeueReusableCell(withIdentifier: "cell", for: indexPath)
+        cell.selectionStyle = .none
+        cell.backgroundColor = .ypLigthGray.withAlphaComponent(0.3)
+        cell.textLabel?.text = viewModel.categoriesArray[indexPath.row]
```
Старайся не работать со свойствами ячейки и вообще класса вот так напрямую из другого класса. 
По хорошему - это сделать в ячейке публичную функцию, например configCell с аргументами на вход, и уже в ней менять нужные свойства. 
Свойства класса нужно всегда стараться закрывать извне, и работать с ними только изнутри

andrew-kuzn-yndx commented on Tracker/Controllers/CategoryViewController/CategoryListViewController.swift:
```diff
@@ -0,0 +1,275 @@
+            }
+            return UIMenu(title: "", children: [editAction, deleteAction])
+        }
+        return configuration
```
Этот метод очень нагруженный и плохо влияет на производительность
Давай разгрузим его) например показ алертов можно вывести в отдельные вспомогательные методы

andrew-kuzn-yndx commented on Tracker/Controllers/FiltersViewController/FiltersViewController.swift:
```diff
@@ -0,0 +1,118 @@
+            self.selectedFilter = filtersArray[indexPath.row]
+            let checkmarkImageView = UIImageView(image: UIImage(named: "checkmark"))
+            tableView.cellForRow(at: indexPath)?.accessoryView = checkmarkImageView
+            UserDefaults.standard.set(self.selectedFilter?.rawValue, forKey: "selectedFilter")
```
давай разгрузим все методы tableView во всем проекте, Что-то выведем в отдельные методы, что-то спрячем внутрь ячеек

andrew-kuzn-yndx commented on Tracker/Controllers/TrackersViewController/TrackersCell.swift:
```diff
@@ -2,23 +2,18 @@ import UIKit
     private var indexPath: IndexPath?
     
-    private lazy var backView: UIView = {
+    lazy var backView: UIView = {
```
Лучше сделать свойства обратно приватным, и менять его изнутри, а не снаружи 
тоже самое про counterLabel

andrew-kuzn-yndx commented on Tracker/Models/Weekdays.swift:
```diff
@@ -1,49 +1,70 @@
-    case Friday = "Пятница"
-    case Saturday = "Суббота"
-    case Sunday = "Воскресенье"
+    case Monday = "Monday"
```
не забывай, что кейсы энама мы именуем с мальенькой буквы)

Commit:
Исправлены замечания:
1) Исправлено: При смене дней недели, у тебя показывается заглушка (плейсхолдер) предназначенный для пустых результатов поиска.
2) Исправлено: После создания трекера и закрытия этого экрана, открывается пустой экран трекеров без заглушки
3) Исправлено: Что-то странное происходит с созданием и выбором категорий
У меня все нормально: залил видео - https://disk.yandex.ru/i/nkGqRWWxpt4iLg
Также исправлены другие замечания из комментариев к ПР

Commit:
Удалены принты
